darwin/dataset/upload_manager.py

```diff
@@ -395,6 +452,99 @@ def callback(monitor):
+            )
+
+        file_lookup = {file.full_path: file for file in self.local_files}
+        for item in self.pending_items:
+            if len(item.slots) != 1:
+                raise NotImplemented("Multi file upload is not supported")
+            upload_id = item.slots[0]["upload_id"]
+            file = file_lookup.get(item.full_path)
+            if not file:
+                raise ValueError(f"Cannot match {item.full_path} from payload with files to upload")
+            yield upload_function(self.dataset.identifier.dataset_slug, file.local_path, upload_id)
+
+    def _upload_file(
+        self, dataset_slug: str, file_path: Path, upload_id: str, byte_read_callback: Optional[ByteReadCallback]
+    ) -> None:
+        try:
+            self._do_upload_file(dataset_slug, file_path, upload_id, byte_read_callback)
+        except UploadRequestError as e:
+            self.errors.append(e)
+        except Exception as e:
+            self.errors.append(UploadRequestError(file_path=file_path, stage=UploadStage.OTHER, error=e))
+
+    def _do_upload_file(
+        self,
+        dataset_slug: str,
+        file_path: Path,
+        upload_id: str,
+        byte_read_callback: Optional[ByteReadCallback] = None,
+    ) -> None:
+        team_slug: Optional[str] = self.dataset_identifier.team_slug
+
+        try:
+            sign_response: Dict[str, Any] = self.client.api_v2.sign_upload(dataset_slug, upload_id, team_slug=team_slug)
+        except Exception as e:
+            raise UploadRequestError(file_path=file_path, stage=UploadStage.REQUEST_SIGNATURE, error=e)
+
+        upload_url = sign_response["upload_url"]
+
+        try:
+            file_size = file_path.stat().st_size
+            if byte_read_callback:
+                byte_read_callback(str(file_path), file_size, 0)
+
+            def callback(monitor):
+                if byte_read_callback:
+                    byte_read_callback(str(file_path), file_size, monitor.bytes_read)
+
+            with file_path.open("rb") as m:
+                monitor = FileMonitor(m, file_size, callback)
+
+                retries = 0
+                while retries < 5:
+                    upload_response = requests.put(f"{upload_url}", data=monitor)
+                    # If s3 is getting to many request it will return 503, we will sleep and retry
+                    if upload_response.status_code != 503:
+                        break
+
+                    time.sleep(2 ** retries)
+                    retries += 1
+
+            upload_response.raise_for_status()
+        except Exception as e:
+            raise UploadRequestError(file_path=file_path, stage=UploadStage.UPLOAD_TO_S3, error=e)
+
+        try:
+            self.client.api_v2.confirm_upload(dataset_slug, upload_id, team_slug=team_slug)
+        except Exception as e:
+            raise UploadRequestError(file_path=file_path, stage=UploadStage.CONFIRM_UPLOAD_COMPLETE, error=e)
+
```

Comment:
As far as I can see, all that changed here is that we point to `api_v2`. If that's the case, looks good. If not, the code is quite hard to parse, so would be great to be pointed to the parts of the code that substantially differ from v1.